substrate-node/pallets/pallet-dao/src/dao.rs

```diff
@@ -0,0 +1,366 @@
+            Error::<T>::WrongIndex
+        );
+
+        // Don't allow votes after the time limit reached
+        ensure!(
+            frame_system::Pallet::<T>::block_number() <= voting.end,
+            Error::<T>::TimeLimitReached
+        );
+
+        let position_yes = voting.ayes.iter().position(|a| a.farm_id == farm_id);
+        let position_no = voting.nays.iter().position(|a| a.farm_id == farm_id);
+
+        // Detects first vote of the member in the motion
+        let is_account_voting_first_time = position_yes.is_none() && position_no.is_none();
+
+        if approve {
+            if position_yes.is_none() {
+                voting.ayes.push(proposal::VoteWeight {
+                    farm_id: farm_id,
+                    weight: Self::get_vote_weight(farm_id)?,
+                });
+            } else {
+                return Err(Error::<T>::DuplicateVote.into());
+            }
+            if let Some(pos) = position_no {
+                voting.nays.swap_remove(pos);
+            }
+        } else {
+            if position_no.is_none() {
+                voting.nays.push(proposal::VoteWeight {
+                    farm_id: farm_id,
+                    weight: Self::get_vote_weight(farm_id)?,
+                });
+            } else {
+                return Err(Error::<T>::DuplicateVote.into());
+            }
+            if let Some(pos) = position_yes {
+                voting.ayes.swap_remove(pos);
+            }
+        }
+
+        let yes_votes = voting.ayes.len() as u32;
+        let no_votes = voting.nays.len() as u32;
+        Self::deposit_event(Event::Voted {
+            account: who,
+            proposal_hash,
+            voted: approve,
+            yes: yes_votes,
+            no: no_votes,
+        });
+
+        Voting::<T>::insert(&proposal_hash, voting);
+
+        if is_account_voting_first_time {
+            Ok(Pays::No.into())
+        } else {
+            Ok(Pays::Yes.into())
+        }
+    }
+    pub fn _veto(who: T::AccountId, proposal_hash: T::Hash) -> DispatchResultWithPostInfo {
+        Self::is_council_member(who.clone())?;
+
+        let stored_proposal =
+            <Proposals<T>>::get(proposal_hash).ok_or(Error::<T>::ProposalMissing)?;
+
+        let mut voting = Self::voting(proposal_hash).ok_or(Error::<T>::ProposalMissing)?;
+        ensure!(
+            voting.index == stored_proposal.index,
+            Error::<T>::WrongIndex
+        );
+
+        // push vote to vetos
+        voting.vetos.push(who.clone());
+
+        Self::deposit_event(Event::CouncilMemberVeto { proposal_hash, who });
+
+        if voting.vetos.len() as u32 >= T::MinVetos::get() {
+            Self::deposit_event(Event::ClosedByCouncil {
+                proposal_hash,
+                vetos: voting.vetos,
+            });
+            Self::do_disapprove_proposal(proposal_hash);
+            return Ok(Pays::No.into());
+        }
+
+        Voting::<T>::insert(&proposal_hash, voting);
```

Comment:
The storage is modified after depositing the event. Is that correct? What if someone wants to check the voting right after he sees the CouncilMemberVeto event? He would still stee the old voting.

Reply:
I think the requested changes here are outside the scope of this PR / unrelated to the modules structures improvements incorporated here and it's better to have a separate task and pull request for that.